common/config/chaintype.go

```diff
@@ -6,45 +6,91 @@ import (
 )
 
 // ChainType denotes the chain or network to work with
-type ChainType string
+type ChainType int
```

Comment:
How is this parsed from TOML now? The TOML lib doesn't know about `func ChainTypeFromSlug`. Normally custom types must implement `TextMarshaler` and `TextUnmarshaler`.

Reply:
The TOML parsing hasn't changed, it's still going to be a `*string` after unmarshalling:
https://github.com/smartcontractkit/chainlink/blob/SHIP-2001/decouple-chaintype/core/chains/evm/config/toml/config.go#L351

Only when actually using it, I've changed all conversions `chainType := config.ChainType(*c.ChainType)` to use the newly added function `chainType := config.ChainTypeFromSlug(*c.ChainType)`.

Reply:
But I see your point, cleaner would probably be to add custom unmarshalling and directly add the `ChainType` type to the structs. This would eliminate the confusion of having to convert between string/ChainType

Reply:
Embedded the type directly into the config struct here: https://github.com/smartcontractkit/chainlink/pull/13272/files#diff-7cc03cd5adaef6b2b6e25f221cf16163cbd44f0b8f2bf188febaf01389a7bf9eR344
The workaround with the second type is needed to preserve the original value, making it available for deprecation checks